File: android/appmanager/src/test/java/com/lmig/uscm/digital/appplatform/appmanager/FeatureManagerTests.kt
package com.lmig.uscm.digital.appplatform.appmanager

import android.content.Context
import android.content.Intent
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.*
import java.lang.reflect.Type
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * FeatureManager Tests
 */
class FeatureManagerTests {

    private var featureMan: FeatureManager? = null

    var mockContext: Context? = null

    class FeatureTest : Feature {
        var setupCalled = false
        private val mockOptionsJson = "{\"name\":\"Default\",\"config\":{\"theme\":\"light\"},\"viewConfig\":{\"name\":\"ReactNative\",\"config\":{\"bundleURL\":\"http://localhost:8081/index.ios.bundle?platform=ios\",\"moduleName\":\"DefaultModule\",\"initialProperties\":{\"hello\":\"I'm an initial property\"},\"launchOptions\":null}}}"
        private val typeOfHashMap: Type = object : TypeToken<Map<String, Any>>() {}.type
        private val gson: Gson = GsonBuilder().serializeNulls().create()
        private val options: Map<String, Any> = gson.fromJson(mockOptionsJson, typeOfHashMap)

        override val name = "Test"
        override var config: FeatureConfig? = FeatureConfig(options)
        override val intent: Intent = mock(Intent::class.java)

        override fun setup() {
            setupCalled = true
        }
    }

    var feature: FeatureTest? = null

    @Before
    fun setUp() {
        mockContext = mock(Context::class.java, RETURNS_DEEP_STUBS)
        doNothing().`when`(mockContext)?.startActivity(any())
        feature = FeatureTest()
        feature?.let { feature ->
            mockContext?.let { mockContext ->
                featureMan = FeatureManager(features = arrayListOf(feature),defaultFeatureName = "Test", context = mockContext)
            }
        }
    }

    @After
    fun tearDown() {
    }

    @Test
    fun testInit() {
        featureMan?.let { featureMan ->
            assertEquals(1, featureMan.features.size)
            assertEquals("Test", featureMan.defaultFeatureName)
            assertEquals("Test", featureMan.currentFeatureName)
            assertTrue(featureMan.context is Context)
        }
    }

    @Test
    fun testUpdateFeature() {
        featureMan?.let { featureMan ->
            feature?.let { feature ->
                val mockOptionsJson = "{\"moduleName\":\"NewNav\",\"initialProperties\":{\"initProp1\":\"initProp1NewValue\"},\"launchOptions\":{\"launchOption1\":\"launchOption1NewValue\"}}"
                val typeOfHashMap: Type = object : TypeToken<Map<String, Any>>() {}.type
                val gson: Gson = GsonBuilder().serializeNulls().create()
                val options: Map<String, Any> = gson.fromJson(mockOptionsJson, typeOfHashMap)
                val reactNav = ReactNav(attributes = options)
                val nav = Navigation(featureName = "TEST", data = reactNav)

                featureMan.updateFeature(feature = feature, nav = nav)

                val rnvc = feature.config?.viewConfig?.config as ReactNativeViewConfig

                // Test that a ReactNativeViewConfig was created
                assertTrue(feature.config?.viewConfig?.config is ReactNativeViewConfig)

                // Test that view config module name has been populated correctly
                assertEquals("NewNav", rnvc.moduleName)

                // Test that view config initial properties have been populated correctly
                assertEquals("initProp1NewValue", rnvc.initialProperties?.get("initProp1"))

                // Test that view config launch options have been populated correctly
                assertEquals("launchOption1NewValue", rnvc.launchOptions?.get("launchOption1"))

                // Test that the setup() function on the feature has been called
                assertTrue(feature.setupCalled)

            }
        }
    }

    @Test
    fun testDisplay() {
        featureMan?.let { featureMan ->
            feature?.let { feature ->
                featureMan.display(feature = feature)
            }
        }
        verify(mockContext)?.startActivity(any())
    }

    @Test
    fun testDisplayWithAnimation() {
        featureMan?.let { featureMan ->
            feature?.let { feature ->
                featureMan.displayWithAnimation(feature = feature, animation = "Test")
            }
        }
        verify(mockContext)?.startActivity(any())
    }

    @Test
    fun testMergeConfig() {
        featureMan?.let { featureMan ->
            val mockOptionsJson = "{\"name\":\"Test\",\"viewConfig\":{\"name\":\"Test\",\"config\":{\"name\":\"Test\"}}}"
            val typeOfHashMap: Type = object : TypeToken<Map<String, Any>>() {}.type
            val gson: Gson = GsonBuilder().serializeNulls().create()
            val options: Map<String, Any> = gson.fromJson(mockOptionsJson, typeOfHashMap)

            val featureConf = FeatureConfig(attributes = options)
            featureMan.mergeConfig(featureConfig = arrayListOf(featureConf))

            feature?.let { feature ->

                // Test the feature has been updated with the new feature config
                assertEquals(featureConf, feature.config)
            }
        }
    }

    @Test
    fun testSetupViews() {
        featureMan?.let { featureMan ->
            val mockOptionsJson = "{\"name\":\"Default\",\"config\":{\"moduleName\":\"TestModule\",\"initialProperties\":{\"initProp1\":\"initProp1Value\"},\"launchOptions\":{\"launchOption1\":\"launchOption1Value\"},\"bundleURL\":\"https://www.libertymutual.com\"}}"
            val typeOfHashMap: Type = object : TypeToken<Map<String, Any>>() {}.type
            val gson: Gson = GsonBuilder().serializeNulls().create()
            val options: Map<String, Any> = gson.fromJson(mockOptionsJson, typeOfHashMap)

            val viewConf = ViewConfig(attributes = options)
            featureMan.setupViews(viewConfig = arrayListOf(viewConf))

            feature?.let { feature ->
                // Test the feature has been updated with the new view config
                assertEquals(viewConf, feature.config?.viewConfig)
            }
        }
    }

    @Test
    fun testGetDefaultFeature() {
        featureMan?.let{
            assertEquals(feature, it.getDefaultFeature())
        }
    }

    @Test
    fun testGetFeatureByName() {
        featureMan?.let{
            assertEquals(feature, it.getFeatureByName(name = "Test"))
        }
    }
}

File: ios/AppManagerTests/FeatureManagerTests.swift
//
//  FeatureManagerTests.swift
//  AppManager
//
//  Copyright © 2017 Liberty Mutual. All rights reserved.
//

import XCTest
import AppPlatform

class FeatureManagerTests: XCTestCase {
    var feature: TestFeature?
    let attributes = [
        "name": "Test",
        "config": ["theme": "light"],
        "viewConfig": [
            "name": "ReactNative",
            "config": [
                "bundleURL": "http://localhost:8081/index.ios.bundle?platform=ios",
                "moduleName": "HelloWorld",
                "initialProperties": nil,
                "launchOptions": nil
            ]
        ]
        ] as [String: Any]

    override func setUp() {
        super.setUp()

        if let featureConfig = try? FeatureConfig(attributes: self.attributes) {
            self.feature = TestFeature(config: featureConfig)
        }
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testInit() {
        let manager = FeatureManager(feature: self.feature!)
        XCTAssertEqual(manager.defaultFeature, self.feature?.name)

        if let f = manager.features.first as? TestFeature {
            XCTAssertEqual(f.name, self.feature?.name)
        } else {
          XCTFail("Not a TestFeature")
        }
    }

    func testInitMultiple() {
        let manager = FeatureManager(features: [self.feature!], defaultFeature: "Test")
        XCTAssertEqual(manager.defaultFeature, self.feature?.name)

        if let f = manager.features.first as? TestFeature {
            XCTAssertEqual(f.name, self.feature?.name)
        } else {
            XCTFail("Not a TestFeature")
        }
    }

    func testDisplay() {
        let manager = FeatureManager(feature: self.feature!)

        // Not testable at the moment
        manager.display(feature: self.feature!)
        XCTAssertNil(nil)

    }

    func testMergeConfig() {
        let manager = FeatureManager(feature: self.feature!)
        manager.mergeConfig(featureConfig: [self.feature!.config!])

        XCTAssert(manager.features.first!.config == self.feature!.config)
    }

    func testMergeConfigEmpty() {
        let manager = FeatureManager(feature: self.feature!)
        manager.features.first!.config = nil
        if let featureConfig = try? FeatureConfig(attributes: self.attributes) {
//            self.feature = TestFeature(config: featureConfig)
            manager.mergeConfig(featureConfig: [featureConfig])

            XCTAssert(manager.features.first!.config == self.feature!.config)
        } else {
            XCTFail("FeatureConfig init failed")
        }
    }

    func testSetupViews() {
        let attributes = [
            "name": "ReactNative",
            "config": [
                "bundleURL": "http://localhost:8081/index.ios.bundle?platform=ios",
                "moduleName": "HelloWorld",
                "initialProperties": nil,
                "launchOptions": nil
            ]
            ] as [String: Any]

        let manager = FeatureManager(feature: self.feature!)
        if let viewConfig = try? ViewConfig(attributes: attributes) {
            manager.setupViews(viewConfig: [viewConfig])
            XCTAssert(manager.features.first?.config?.viewConfig == viewConfig)
        } else {
            XCTFail("ViewConfig not initialized")
        }
    }

    func testSetupViewsEmptyConfig() {
        let attributes = [
            "name": "ReactNative",
            "config": [
                "bundleURL": "http://localhost:8081/index.ios.bundle?platform=ios",
                "moduleName": "HelloWorld",
                "initialProperties": nil,
                "launchOptions": nil
            ]
            ] as [String: Any]

        let manager = FeatureManager(feature: self.feature!)
        if let viewConfig = try? ViewConfig(attributes: attributes) {
            manager.features.first?.config?.viewConfig = nil
            manager.setupViews(viewConfig: [viewConfig]) // Quiet failure (print only)
            XCTAssertNil(manager.features.first?.config?.viewConfig)
        } else {
            XCTFail("ViewConfig not initialized")
        }
    }

    func testGetDefaultFeature() {
        let manager = FeatureManager(feature: self.feature!)
        let defaultFeature = manager.getDefaultFeature() as? TestFeature
        XCTAssertEqual(defaultFeature?.config, self.feature!.config)
        XCTAssertEqual(defaultFeature?.name, self.feature!.name)

        manager.features = []
        XCTAssertNil(manager.getDefaultFeature())
    }

    func testUpdateFeature() {
        let manager = FeatureManager(feature: self.feature!)
        let attributes = [
            "moduleName": "HelloWorld",
            "initialProperties": ["testing": 1],
            "launchOptions": ["launch": "option"]
        ] as [String: Any]
        if let reactNativeData = try? ReactNav(attributes: attributes) {
            let nav = Navigation(featureName: "Test", data: reactNativeData)
            manager.updateFeature(feature: self.feature!, nav: nav)
            let rnConf = manager.getDefaultFeature()?.config?.viewConfig?.config as? ReactNativeViewConfig
            XCTAssertEqual((rnConf?.initialProperties as? [String: Int])!, ["testing": 1])
            XCTAssertEqual((rnConf?.launchOptions as? [String: String])!, ["launch": "option"])
        } else {
            XCTFail("ReactNav init failed")
        }
    }

    func testGetFeatureByName() {
        let manager = FeatureManager(feature: self.feature!)
        XCTAssertEqual(manager.getFeatureByName(name: "Test")!.name, self.feature!.name)
        XCTAssert(manager.getFeatureByName(name: "TestNonExistant") == nil)
    }

    func testGetFeatureByNameFake() {
        let manager = FeatureManager(feature: self.feature!)
        manager.features = []
        XCTAssertNil(manager.getFeatureByName(name: ""))
    }

}
